Allow deleting assets

SaraVieira opened the pull request:
This will delete an image from both teh DB and the file system/s3

related https://github.com/webstudio-is/webstudio-designer/issues/37

## Before requesting a review

- [ ] if the PR is WIP - use draft mode
- [x] made a self-review
- [x] added inline comments where things may be not obvious (the "why", not "what")
- [ ] what kind of review is needed?
  - [x] conceptual
  - [ ] detailed
  - [x] with testing

## Test cases

- [ ] step by step interaction description and what is expected to happen

## Before merging

- [x] tested locally and on preview environment (preview dev login: 5de6e)
- [x] updated [test cases](https://github.com/webstudio-is/webstudio-designer/blob/main/apps/designer/docs/test-cases.md) document
- [x] added tests
- [x] if any new env variables are added, added them to `.env.example` and the `designer/env-check.js` if mandatory

kof commented on apps/designer/app/routes/designer/$id.tsx:
```diff
@@ -35,17 +38,42 @@ type Error = {
+        const deletedAsset = await deleteAsset({
+          id,
+          name,
+          dirname: __dirname,
```
please don't use it, we need something importable from a single location for __dirname, I feel like I mentioned it before and wonder now if we have multiple places now

SaraVieira replied:
Then i can't delete the file locally. Do you have a better option?

kof commented on apps/designer/app/routes/designer/$id.tsx:
```diff
@@ -35,17 +38,42 @@ type Error = {
+      const formData = await request.formData();
+      if (formData.get("assetId")) {
+        const id = formData.get("assetId") as string;
+        const name = formData.get("assetName") as string;
```
instead of asserting with typescript, validate mit zod

SaraVieira replied:
Will do

SaraVieira replied:
Mit zot, went all German there ahahah

SaraVieira replied:
Done with zod-form-data, updated other places to use it

kof commented on apps/designer/app/shared/design-system/stitches.config.ts:
```diff
@@ -219,6 +219,7 @@ export const {
       7: "45px",
       8: "65px",
       9: "80px",
+      10: "105px",
```
please check with @taylornowotny and/or @thysultan regarding the grid we have in the design system

SaraVieira replied:
Will check the grid and ask if i don't find it

SaraVieira replied:
According to the design system everything should actually be in increments of 8, I updated it

kof commented on packages/asset-uploader/src/delete.ts:
```diff
@@ -0,0 +1,41 @@
+      return await deleteAssetInDb(id);
+    }
+  }
+  throw new Error("Asset does not exist");
```
can we throw early? like early return, easier to skip code I don't need to read

SaraVieira replied:
done!|

kof reviewed (changes requested):
found a few things, will test soon

kof commented on apps/designer/app/designer/features/sidebar-left/panels/asset-manager.tsx:
```diff
@@ -2,48 +2,53 @@ import { ImageIcon } from "~/shared/icons";
   useEffect(() => {
-    if (newImages?.length) {
+    if (imageChanges?.uploadedAssets?.length) {
       setAssets((currentAssets) => [
```
I wonder if there was an actual case for using setState callback here

SaraVieira replied:
hm?

kof replied:
setAsset(callback) vs assets

SaraVieira replied:
I always tend to use the callback, its way safer

kof replied:
by doing so you are waiting for the next update cycle, this is only useful when you know there can be an update that came from somewhere else, its not a thing we need to use as a general pattern, only in special cases

SaraVieira replied:
removing it actually broke it, this makes sure we are using the latest version

kof replied:
because you didn't set the dependency?

SaraVieira replied:
I did,updated it to include baseAssets

kof commented on apps/designer/app/designer/features/sidebar-left/panels/asset-manager.tsx:
```diff
@@ -2,48 +2,53 @@ import { ImageIcon } from "~/shared/icons";
-            alt={asset.alt || asset.name}
-            status={asset.status}
-          />
+          <AssetManagerImage key={asset.id} {...asset} />
```
This component is so much better off with a "Thumbnail" as a name :)

SaraVieira replied:
updated!

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/image.tsx:
```diff
@@ -18,18 +26,17 @@ const useImageWithFallback = ({ path }: { path: string }) => {
+  useClickAway(tooltipRef, () => {
+    setTolltipOpen(false);
+  });
+  useKey("Escape", () => setTolltipOpen(false));
```
I wonder if we should use consistently useHotkeys for this

SaraVieira replied:
definitly, updated

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/image.tsx:
```diff
@@ -67,6 +81,37 @@ export const AssetManagerImage = ({
       ></Box>
+      {!isUploading && (
+        <Box ref={tooltipRef}>
+          <Tooltip
```
is this potentially how our default tooltip inside design system should always be styled?

SaraVieira replied:
yeah, I agree, gonna change it in the tooltip

kof commented:
I wanted to test on preview this PR, is this broken like this on main or only this branch?

https://webstudio-designer-g0pu1k7il-webstudio-is.vercel.app/login
<img width="1430" alt="Screenshot 2022-07-26 at 20 01 16" src="https://user-images.githubusercontent.com/52824/181078650-8790e2cb-2312-4957-ac16-383f577cf1bf.png">

kof commented:
What's the keyboard accessibility story here with asset manager?

kof commented:
<img width="581" alt="Screenshot 2022-07-26 at 20 05 52" src="https://user-images.githubusercontent.com/52824/181079556-f4bc9a28-88cd-4e7d-91fe-a9720da425ee.png">

kof commented:
cc @taylornowotny 

<img width="273" alt="Screenshot 2022-07-26 at 20 14 09" src="https://user-images.githubusercontent.com/52824/181081130-fcf62742-04ac-4bdc-a725-03ebf57c7de4.png">

kof commented:
@SaraVieira delete button click doesn't do anything

SaraVieira commented:
@kof the issue with login was because I cahnged the spacing everywhere. Needs to be in a spearate PR

Reverted and added the width in px and added a todo to the tooltip

SaraVieira commented:
@kof fixed the button, it was the Box that was around the tooltip was not catching anything.

We can now pass a ref to the tooltip and I also added the loading progress bar to the deleting to be more obvious what is happening

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/thumbnail.tsx:
```diff
@@ -0,0 +1,135 @@
+  const submit = useSubmit();
+  const isUploading = status === "uploading";
+  const src = useImageWithFallback({ path });
+  const [isToolTipOpen, setTolltipOpen] = useState(false);
```
```suggestion
  const [isTooltipOpen, setTolltipOpen] = useState(false);
```

SaraVieira replied:
Fixed

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/thumbnail.tsx:
```diff
@@ -0,0 +1,135 @@
+
+  const tooltipRef = useRef(null);
+  useClickAway(tooltipRef, closeTooltip);
+  useHotkeys(shortcuts["escape"], closeTooltip);
```
radix tooltip is supposed to handle both, esc and click outside https://www.radix-ui.com/docs/primitives/components/tooltip#content

SaraVieira replied:
I see what happened, it had to be updated. Updated it to v1 and it now supports everything

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/thumbnail.tsx:
```diff
@@ -18,29 +26,41 @@ const useImageWithFallback = ({ path }: { path: string }) => {
       );
     },
-    isUploading ? 100 : null
+    isUploading || isDeleting ? 100 : null
```
why is it even taking null, would have expected 0 or undefined there

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/thumbnail.tsx:
```diff
@@ -18,29 +26,41 @@ const useImageWithFallback = ({ path }: { path: string }) => {
+  const closeTooltip = () => setTolltipOpen(false);
+
   // @todo rewrite this fake indication to show real progress
   useInterval(
```
would move it into useFakeProgress

SaraVieira replied:
Done!

kof commented on apps/designer/app/routes/designer/$id.tsx:
```diff
@@ -33,19 +37,46 @@ type Error = {
+      const deletedAsset = await deleteAsset({
+        id: assetId,
+        name: assetName,
+        dirname: __dirname,
```
tbh I don't even understand why we need __dirname, which is the routes/designer in this case ... I would need to follow this code through but on the surface it doesn't make sense, because that folder where we handle the route should not be used at all

kof commented on apps/designer/package.json:
```diff
@@ -43,7 +43,7 @@
     "@radix-ui/react-toggle": "^0.1.4",
     "@radix-ui/react-toggle-group": "^0.1.5",
-    "@radix-ui/react-tooltip": "^0.1.7",
+    "@radix-ui/react-tooltip": "^1.0.0",
```
we should probably try to upgrade them all unless it breaks something) also feel free to separate the upgrade

SaraVieira replied:
yeah, should definitely be another PR

SaraVieira replied:
tried it and unfortunatly we need to see why it all errors out

![image](https://user-images.githubusercontent.com/1051509/181220670-7c5bbda2-fdc4-4a7d-81ec-821d754fd6b6.png)

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/thumbnail.tsx:
```diff
@@ -0,0 +1,142 @@
+  return src;
+};
+
+export const useFakeProgress = ({
```
why do you export

SaraVieira replied:
Removed, its habit

kof commented on packages/asset-uploader/src/constants.ts:
```diff
@@ -1 +1 @@
-export const DEFAULT_UPLPOAD_PATH = "uploads";
+export const DEFAULT_UPLPOAD_PATH = "pub lic/uploads";
```
`pub lic`???

kof commented on apps/designer/app/designer/features/sidebar-left/panels/components/thumbnail.tsx:
```diff
@@ -0,0 +1,142 @@
+
+  return (
+    <Box
+      title={alt || ""}
```
some better fallback title maybe?

kof replied:
like a meaningful one

SaraVieira replied:
added the name

kof commented on packages/asset-uploader/src/delete.ts:
```diff
@@ -0,0 +1,38 @@
+  id: string;
+  name: string;
+}): Promise<Asset> => {
+  const currentAsset = await prisma?.asset.findUnique({
```
how is prisma potentially not defined?
```suggestion
  const currentAsset = await prisma.asset.findUnique({
```

SaraVieira replied:
done

kof commented on packages/asset-uploader/src/delete.ts:
```diff
@@ -0,0 +1,38 @@
+
+    return await deleteAssetInDb(id);
+  } else {
+    const directory = await imageFSDirectory();
```
```suggestion
    const directory = await imageFsDirectory();
```

SaraVieira replied:
done

kof commented on packages/asset-uploader/src/helpers/image-fs-path.ts:
```diff
@@ -0,0 +1,18 @@
+export const imageFSDirectory = () => {
+  const basePath =
+    process.env.NODE_ENV === "development"
+      ? path.join(
```
please avoid ternaries when you can, this is a perfect example where you could have returned early instead

SaraVieira replied:
I am not returning early, I am deploying this to add the other part of the if

kof replied:
but you can

SaraVieira replied:
i was still testing, it's changed now

I needed to add the version for prod

kof reviewed (approved):
see a bunch of nit picks and lets go
